src/irmin-pack/inode.ml

```diff
@@ -1444,13 +1511,128 @@ struct
                 | None -> false
                 | Some t ->
                     let expected_depth = inodes.depth + 1 in
-                    Ptr.target ~expected_depth ~cache:true layout t
+                    Ptr.target ~expected_depth ~cache:true ~force:true
+                      "contains_empty_map" layout t
                     |> check_lower)
               inodes.entries
       in
       check_lower t
 
     let is_tree t = match t.v with Tree _ -> true | Values _ -> false
+
+    module Proof = struct
+      type value = [ `Contents of hash * metadata | `Node of hash ]
+      [@@deriving irmin]
+
+      type t =
+        [ `Blinded of hash
+        | `Values of (step * value) list
+        | `Inode of int * (int * t) list ]
+      [@@deriving irmin]
+
+      let weaken_step_value (step, v) = (step, hashvalue_of_keyvalue v)
+
+      let strengthen_step_value (step, v) =
+        (* Since proofs only in the context of portable, using this unsafe
+           function is safe. *)
+        (step, unsafe_keyvalue_of_hashvalue v)
+
+      let rec proof_of_concrete :
+          type a. hash Lazy.t -> Concrete.t -> (t -> a) -> a =
+       fun h proof k ->
+        match proof with
+        | Blinded -> k (`Blinded (Lazy.force h))
+        | Values vs ->
+            let l =
+              List.map Concrete.of_entry vs |> List.map weaken_step_value
+            in
+            k (`Values l)
+        | Tree tr ->
+            let proofs =
+              List.fold_left
+                (fun acc (e : _ Concrete.pointer) ->
+                  let hash = lazy e.pointer in
+                  proof_of_concrete hash e.tree (fun proof ->
+                      (e.index, proof) :: acc))
+                [] (List.rev tr.pointers)
+            in
+            k (`Inode (tr.length, proofs))
+
+      let hash_values ~depth l =
+        let t =
+          match depth with
+          | 0 -> of_seq Truncated (List.to_seq l)
```

Comment:
maybe the code is a bit clearer if we add a `values_root` function? (instead of going through `of_seq`)?